src/main/scala/com/databricks/labs/overwatch/pipeline/Snapshot.scala

```diff
@@ -0,0 +1,308 @@
+    val acceptableCloneLevels = Array("DEEP", "SHALLOW")
+    require(acceptableCloneLevels.contains(cloneLevel.toUpperCase), s"SNAP CLONE ERROR: cloneLevel provided is " +
+      s"$cloneLevel. CloneLevels supported are ${acceptableCloneLevels.mkString(",")}.")
+
+    val sourceToSnap = pipelineTables
+    val exclude = excludes match {
+      case Some(s) if s.nonEmpty => s
+      case _ => ""
+    }
+    val excludeList = exclude.split(",")
+
+    val cleanExcludes = excludeList.map(_.toLowerCase).map(exclude => {
+      if (exclude.contains(".")) exclude.split("\\.").takeRight(1).head else exclude
+    })
+
+
+    val sourceToSnapFiltered = sourceToSnap
+      .filter(_.exists()) // source path must exist
+      .filterNot(t => cleanExcludes.contains(t.name.toLowerCase))
+
+    val cloneSpecs = buildCloneSpecs(cloneLevel,sourceToSnapFiltered)
+    val cloneReport = Helpers.parClone(cloneSpecs)
+    val cloneReportPath = s"${snapshotRootPath}/clone_report/"
+    cloneReport.toDS.write.format("delta").mode("append").save(cloneReportPath)
+    this
+  }
+
+}
+
+object Snapshot extends SparkSessionWrapper {
+
+  private val logger: Logger = Logger.getLogger(this.getClass)
+
+  def apply(workspace: Workspace,
+            sourceETLDB : String,
+            targetPrefix : String,
+            snapshotType: String,
+            excludes: Option[String],
+            cloneLevel: String,
+            pipelineTables : Array[PipelineTable],
+            processType : String,
+           ): Snapshot = {
+    if (snapshotType.toLowerCase()== "incremental")
+      new Snapshot(sourceETLDB, targetPrefix, workspace, workspace.database, workspace.getConfig,processType).incrementalSnap(pipelineTables,excludes)
+    else{
+      new Snapshot(sourceETLDB, targetPrefix, workspace, workspace.database, workspace.getConfig,processType).snap(pipelineTables,cloneLevel,excludes)
+    }
+
+  }
+
+
+
+
+  /**
+   * Create a backup of the Overwatch datasets
+   *
+   * @param arg(0)        Source Database Name.
+   * @param arg(1)        Target snapshotRootPath
+   * @param arg(2)        Define the Medallion Layers. Argumnent should be in form of "Bronze, Silver, Gold"(All 3 or any combination of them)
+   * @param arg(3)        Type of Snapshot to be performed. "Full" for Full Snapshot , "Incremental" for Incremental Snapshot
+   * @param arg(4)        Array of table names to exclude from the snapshot
+   *                      this is the table name only - without the database prefix. By Default it is empty.
+   * @param arg(5)        Clone Level for Snapshot. By Default it is "Deep". You can also specify "Shallow" Clone.
+   *
+   * @param arg(6)        This argument specify the process type. Whether it is Restore, Migration or Snapshot. By Default it is Snapshot. This argument is
+   *                      used internally by restore or Migration process by changing this argument.
+   * @return
+   */
+
+  def main(args: Array[String]): Any = {
```

Comment:
this function shouldn't be called main, but instead something else that is more explicit (doSnapshot, startSnapshot, anything else, not main). The params should be named and optional when necessary, but should not be positional args.

Reply:
Done